Feature/#55 feature 좋아요 기능 만들기

jw-jin opened the pull request:
## 개요
Redis 좋아요 기능 적용 및 스케줄링 적용
RedisConfig RedisTemplate 일부 스펙 변경
MemberController Response 수정
RefreshToken 만료시간 수정
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->

<!---- Resolves: #(Isuue Number) -->
Resolves: #55 
## PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [x] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [x] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다. Commit message convention 참고  (Ctrl + 클릭하세요.)
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

jw-jin committed:
Feat: Redis 단순 좋아요 기능 추가
Redis 단순 좋아요 기능 추가
테스트용 oauth yml 추가
테스트 코드 작성

jw-jin committed:
Feat: Add LikesScheduler
좋아요 스케줄러 적용 및 RedisConfig 수정
네이밍 잘못된 클래스들 like -> likes 수정

jw-jin committed:
Feat: LikeController 추가
LikeController 추가
UpdateMemberRequestDto 수정
RefreshToken 만료시간 수정

shinb-bong commented on src/main/java/summer/mrplaylist/likes/service/LikesRedisService.java:
```diff
@@ -0,0 +1,39 @@
+		return setOperations.members(key);
+	}
+
+	public Set<String> getKeys(String pattern) {
```
![image](https://github.com/2023Summer/mr-playlist-backend/assets/81786969/67a2fe25-6702-4cf1-95e9-30080dd00dcd)

등을 보면 지양하기도하고
현재대로하면  만약 지금 검색, 이메일 인증 , 속도가 느리면 데이터 캐싱까지 진행할 예정인데
좋아요와 관련없는 모든 키를 들고오게 되지않을까요?

패턴을 지정해서 서두의 `likes_` 라던지 Custom Prefix를 붙이는게 좋을 듯 합니다.

jw-jin replied:
그 부분은 Likes를 위한 Redis 서버를 따로 관리하는 것으로 하려고 했습니다.
더 찾아보고 수정해볼게요!

jw-jin replied:
최대한 Prefix로 필요한 키만 가져올 수 있도록 분리하였고, 후에 RedisService로 통합하면서 scan 명령까지 적용하겠습니다!

shinb-bong commented on src/main/java/summer/mrplaylist/likes/service/LikesRedisService.java:
```diff
@@ -0,0 +1,39 @@
+		setOperations.remove(key, value);
+	}
+
+	public Set<Long> getAllData(String key) {
```
해당 부분도 아래와 같습니다.

shinb-bong commented on src/main/java/summer/mrplaylist/likes/service/LikesService.java:
```diff
@@ -0,0 +1,35 @@
+	private final LikesRedisService likesRedisService;
+
+	private final LikesRepository likesRepository;
+	private static final String PLAYLIST_LIKES_KEY = "likes:playlist:";
```
변수명도 PREFIX 로 수정이 좋을듯 합니다.

shinb-bong commented on src/main/java/summer/mrplaylist/likes/service/LikesService.java:
```diff
@@ -0,0 +1,35 @@
+		likesRedisService.deleteData(PLAYLIST_LIKES_KEY + playlistId.toString(), memberId);
+	}
+
+	public void playlistLikeSave(Likes likes) {
```
해당 메소드도 스케쥴러에서 사용해서 DB에 삽입할것이라면 `@Transactional` 선언해주세요

shinb-bong commented on src/main/java/summer/mrplaylist/likes/service/LikesRedisService.java:
```diff
@@ -0,0 +1,39 @@
+	@Resource(name = "redisStringLongTemplate")
+	private SetOperations<String, Long> setOperations;
+
+	public void setData(String key, Long value) {
```
여기 addData가 좋을듯합니다.

shinb-bong commented on src/main/java/summer/mrplaylist/likes/service/LikesRedisService.java:
```diff
@@ -0,0 +1,39 @@
+		setOperations.add(key, value);
+	}
+
+	public void deleteData(String key, Long value) {
```
여기도 `removeData` 가 좋을듯합니다.

shinb-bong commented on src/main/java/summer/mrplaylist/likes/service/LikesScheduler.java:
```diff
@@ -0,0 +1,45 @@
+				likesService.playlistLikeSave(likes);
+			}
+		}
+		likesRedisService.flushAll(); // 차후 데이터 백업으로 변경
```
해당 메소드를  사용할 경우 10분뒤에 좋아요를 눌렀던 사용자가 다시 와서 좋아요를 누를 수 있는 상황이 발생할 것 같은데 괜찮을까요....?

아래 메소드에서도 바로 redis만 확인하는 형태로 진행이 되어서 리팩토링이 필요할듯합니다.

jw-jin replied:
급하게 짜느라 부족한 부분이 많네요.. 플로우를 좀더 고민해보고 리팩토링하도록 하겠습니다!

shinb-bong commented on src/test/resources/application.yml:
```diff
@@ -1,6 +1,10 @@
   profiles:
     active: test
+    group:
+      test:
```
`local`로 하고 위에 `spring.profiles.active` 옵션을 local, init 으로 설정 부탁드려요! 
(xlsx 파서랑 test 분리가 의미없어보여서 고쳤습니다!)

shinb-bong reviewed (changes requested):
컨벤션 일치 좋았습니다.

나머지 Message 컨벤션이랑 최종 반환은 한번 회의를 통해 결정하도록하죠!

@ShsH0920 님도 s3가 거의 완성되었다고 하니 다같이 다듬으면 좋을듯합니다.

jw-jin committed:
Refactor: refactor LikesService
Likes 잘못된 변수이름 변경 memberId -> member, playlistId -> playlist
LikesQRepo 추가
LikesService 수정
addLike -> 저장 후 saved로 전환
deleteLike -> redis 우선 접근 후 있으면 redis 좋아요 값 삭제
redis에 없으면 redis DELETE 세팅 후 스케줄러에서 삭제 처리
객체:행동 느낌으로 키 수정 -> LikesConstants 참조

jw-jin committed:
Test: modify test code
테스트 코드 분리
LikesService @Transactional 추가
LikesQRepo @Transactional 추가

jw-jin committed:
Refactor: Add SchedulingConfig
@EnableScheduling Config 화

jw-jin committed:
Merge branch 'develop' into Feature/#55-feature-좋아요-기능-만들기

jw-jin committed:
Rename: 메소드 이름 변경
setData -> addData

jw-jin commented:
@shinb-bong 모두 수정 마쳤습니다. 확인해주시면 감사하겠습니다!

shinb-bong commented on src/main/java/summer/mrplaylist/likes/service/LikesScheduler.java:
```diff
@@ -0,0 +1,71 @@
+	@Scheduled(fixedDelay = 600000) // 10분마다 실행
+	public void updatePlaylistLikes() {
+		Set<String> keys = getLikesKeys();
+		if (keys.isEmpty()) {
```
해당 코드 초기화시에 에러를 출력할 듯합니다.

jw-jin replied:
따로 예외처리 하지 않는 방향으로 가겠습니다!

shinb-bong commented on src/main/java/summer/mrplaylist/likes/repository/LikesQRepo.java:
```diff
@@ -0,0 +1,43 @@
+			.where(
+				qLikes.playlist.id.eq(playlistId)
+					.and(qLikes.member.id.eq(memberId)))
+			.fetchOne();
```
`.fetchFirst()` 로 최적화 하는 것이 좋을 듯합니다. 같은 값 2개 이상시 에러가 납니다.

shinb-bong commented on src/main/java/summer/mrplaylist/likes/repository/LikesQRepo.java:
```diff
@@ -0,0 +1,43 @@
+
+	public boolean existsByPlaylistIdAndMemberId(Long playlistId, Long memberId) {
+		Integer like = queryFactory
+			.selectOne()
```
여기는 어차피 아래서 단건 조회니 select가 좋을듯합니다.

shinb-bong commented on src/main/java/summer/mrplaylist/likes/repository/LikesQRepo.java:
```diff
@@ -0,0 +1,43 @@
+
+@Repository
+@RequiredArgsConstructor
+@Transactional(readOnly = true)
```
여기는 트랜잭션이 불필요합니다.
https://www.inflearn.com/questions/267899/%EB%A0%88%ED%8F%AC%EC%A7%80%ED%86%A0%EB%A6%AC%EC%97%90-transactional%EC%9D%84-%EC%95%88%EB%B6%99%EC%9D%B4%EB%8A%94-%EC%9D%B4%EC%9C%A0

참고자료 첨부합니다.

jw-jin replied:
QRepo 쓰는 서비스의 메소드에 Transactional 처리를 안했어서 문제가 된것 같네요! 서비스에 Transactional 걸도록 하겠습니다!

shinb-bong commented on src/main/java/summer/mrplaylist/likes/controller/LikesController.java:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	@DeleteMapping(DELETE_LIKE)
+	public Response<String> deleteLike(@RequestParam("playlist_id") Long playlistId,
```
해당 컨벤션은 회의에 따라 `@PathVariable`로 일치하겠습니당

shinb-bong reviewed (commented):
반환 타입도 다 일치 시켰고 고생 많으셨습니다!
리뷰 사항 마지막으로 반영 부탁드립니다!

jw-jin committed:
Fix: Fixed LikesService, refactor LikesScheduler
LikesService playlistAddLike 예외 사항 처리 로직 추가
LikesScheduler 예외처리 삭제
LikesQRepo 메소드 쿼리문 수정, @Transactional 삭제 -> 서비스단에 추가
likesRedisService getAllData -> getData 함수명 변경

jw-jin committed:
Merge branch 'Feature/#55-feature-좋아요-기능-만들기' of https://github.com/2023Summer/mr-playlist-backend into Feature/#55-feature-좋아요-기능-만들기
